src/mlir/verona-mlir.cc

```diff
@@ -14,107 +13,111 @@
+    if (inputFile.empty())
+      inputFile = "-";
 
-    try
-    {
-      app.parse(argc, argv);
-    }
-    catch (const CLI::ParseError& e)
+    // Detect source type from extension, if not passed as argument
+    if (inputKind == InputKind::None)
     {
-      exit(app.exit(e));
+      llvm::StringRef filename(inputFile);
+      if (filename.endswith(".verona"))
+        inputKind = InputKind::Verona;
+      else if (filename.endswith(".mlir"))
+        inputKind = InputKind::MLIR;
+      else if (filename == "-") // STDIN, assume Verona
+        inputKind = InputKind::Verona;
     }
 
-    // Default is to output MLIR
-    if (!opt.llvm)
-      opt.mlir = true;
-
-    // Default grammar
-    if (opt.grammar.empty())
-      opt.grammar = path::directory(path::executable()).append("/grammar.peg");
-
-    // Default input is stdin
-    if (opt.filename.empty())
-      opt.filename = "-";
-
-    return opt;
-  }
-
-  // Print help
-  void help()
-  {
-    std::cout << "Compiler Syntax: verona-mlir AST|MLIR|LLVM <filename.verona>"
-              << std::endl;
-  }
-
-  // Detect source type from extension
-  SourceKind getSourceType(llvm::StringRef filename)
-  {
-    auto source = SourceKind::None;
-    if (filename.endswith(".verona"))
-      source = SourceKind::Verona;
-    else if (filename.endswith(".mlir"))
-      source = SourceKind::MLIR;
-    else if (filename == "-") // STDIN, assume MLIR
-      source = SourceKind::MLIR;
-    return source;
-  }
-
-  // Choose output file extension from output type
-  // Careful with mlir->mlir not to overwrite source file
-  std::string
-  getOutputFilename(llvm::StringRef filename, Opt& opt, SourceKind source)
-  {
-    if (!opt.output.empty())
-      return opt.output;
-    if (filename == "-")
-      return "-";
+    // Default to output MLIR
+    if (outputKind == OutputKind::None)
+      outputKind = OutputKind::MLIR;
 
-    std::string newName = filename.substr(0, filename.find_last_of('.')).str();
-    if (opt.mlir)
+    // Choose output file extension from output type
+    // Careful with mlir->mlir not to overwrite source file
+    if (outputFile.empty())
     {
-      if (source == SourceKind::MLIR)
-        newName += ".mlir.out";
+      llvm::StringRef filename(inputFile);
+      if (filename == "-")
+      {
+        outputFile = "-";
+      }
       else
-        newName += ".mlir";
-    }
-    else
-    {
-      newName += ".ll";
+      {
+        std::string newName =
+          filename.substr(0, filename.find_last_of('.')).str();
+        if (outputKind == OutputKind::MLIR)
+        {
+          if (inputKind == InputKind::MLIR)
+            newName += ".new";
+          newName += ".mlir";
+        }
+        else
+        {
+          newName += ".ll";
+        }
+        outputFile = newName;
+      }
     }
-    return newName;
+
+    // Default grammar
+    if (grammarFile.empty())
+      grammarFile = path::directory(path::executable()).append("/grammar.peg");
```

Comment:
What is the use case for allowing the user to override this?  I understand that, for prototyping, we're going to want to be able to quickly iterate on it, but that can be done in the source tree.

Looking for the .peg file in the same location as the executable is also a problem, since it precludes an FHS-compliant installer.

Reply:
`verona-ast` has a case for changing the gramma, `verona-mlir` does not.

However, `verona-mlir` is also a development tool and should never be deployed by an installer.

The actual compiuler will have a driver to resolve all path issues, so this won't be a problem in the actually installable product.